File: javascript/roman-numerals/roman-numerals.js
/*
 * @brief Roman numerals kata.
 */

var ALPHABET =
[
  [1000, "M"], [900, "CM"], [500, "D"], [400, "CD"],
  [ 100, "C"], [ 90, "XC"], [ 50, "L"], [ 40, "XL"],
  [  10, "X"], [  9, "IX"], [  5, "V"], [  4, "IV"],
  [   1, "I"]
];

/**
 * Find the next roman digit
 *
 * @param  {Number} number The arabic number
 * @return {Object}        The value/symbol digit information
 */
function findNextDigit(number)
{
  for (var i = 0; i < ALPHABET.length; ++i)
  {
    digit = ALPHABET[i];

    if (digit[0] <= number)
      return { value: digit[0], symbol: digit[1] };
  }

  return { value: 0, symbol: "" };
}

/**
 * Convert the given number into it's roman string representation
 *
 * @param  {Number} number The number to convert
 * @param  {String} result The result accumulator
 * @return {String}        The roman number representation
 */
function toRoman(number, result)
{
  if (number <= 0)
    return result;

  digit = findNextDigit(number);

  return toRoman(number - digit.value, result.concat(digit.symbol));
}

/**
 * Convert the given number into it's roman string representation
 *
 * @param  {Number} number The number to convert
 * @return {String}        The roman number representation
 */
module.exports.toRoman = function(number)
{
  return toRoman(number, "");
}

File: go/roman-numerals/roman_numerals.go
/**
 * @brief Roman numerals kata.
 */
package roman_numerals

/* IMPORTS *******************************************************************/

import "bytes"

/* STRUCTURES ****************************************************************/

type digit struct {
  value  int
  symbol string
}

/* GLOBALS *******************************************************************/

var alphabet = []digit {
  {1000, "M"}, {900, "CM"}, {500, "D"}, {400, "CD"},
  { 100, "C"}, { 90, "XC"}, { 50, "L"}, { 40, "XL"},
  {  10, "X"}, {  9, "IX"}, {  5, "V"}, {  4, "IV"},
  {   1, "I"},
}

/* PUBLIC FUNCTIONS **********************************************************/

/**
 * Converts the given number into it's roman string representation.
 *
 * @param number The arabic number to convert.
 * @return The roman string representation.
 */
func ToRoman(number int) string {
  var result bytes.Buffer

  for ;number > 0; {
    digit := findNextDigit(number)

    number -= digit.value
    result.WriteString(digit.symbol)
  }

  return result.String()
}

/* PRIVATE FUNCTIONS *********************************************************/

/**
 * Find the next roman digit.
 *
 * @param  number The arabic number
 * @return The value/symbol digit information
 */
func findNextDigit(number int) digit {
  for _, d := range alphabet {
    if d.value <= number {
      return d;
    }
  }

  return digit{ 0, "" }
}
